core/node/node_sync/src/batch_transaction_updater/l1_transaction_verifier.rs

```diff
@@ -0,0 +1,390 @@
+use anyhow::Context as _;
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::{
+    commitment::L1BatchMetadata, ethabi, web3::TransactionReceipt, Address, L1BatchNumber,
+    ProtocolVersionId, H256, U64,
+};
+
+/// Verifies the L1 transaction against the database and the SL.
+#[derive(Debug)]
+pub struct L1TransactionVerifier {
+    diamond_proxy_addr: Address,
+    /// ABI of the ZKsync contract
+    contract: ethabi::Contract,
+    pool: ConnectionPool<Core>,
+    pub validate_logs_from_protocol_version: ProtocolVersionId,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TransactionValidationError {
```

Comment:
Bikeshedding: This looks like an overengineering; AFAICT, errors are not categorized upstream (in `BatchTransactionUpdater`), so `anyhow::Error` with contexts would probably suffice.

Reply:
Will revisit when I work on error handling in `apply_status_update`. I think we need to at least have a separate error for what we want to retry on and what we want to escalate on.